src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/idam/cache/AccessTokenCacheExpiry.java

```diff
@@ -0,0 +1,50 @@
+package uk.gov.hmcts.reform.bulkscan.orchestrator.services.idam.cache;
+
+import com.github.benmanes.caffeine.cache.Expiry;
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import java.util.concurrent.TimeUnit;
+
+@Component
+public class AccessTokenCacheExpiry implements Expiry<String, CachedIdamToken> {
+
+    private final long refreshTokenBeforeExpiry;
+
+    public AccessTokenCacheExpiry(
+        @Value("${idam.client.cache.refresh-before-expire-in-sec}") long refreshTokenBeforeExpiry
+    ) {
+        this.refreshTokenBeforeExpiry = refreshTokenBeforeExpiry;
+    }
+
+    @Override
+    public long expireAfterCreate(
+        @NonNull String key,
+        @NonNull CachedIdamToken tokenResp,
+        long currentTime
+    ) {
+        return TimeUnit.SECONDS.toNanos(tokenResp.expiresIn - refreshTokenBeforeExpiry);
+    }
+
+    @Override
+    public long expireAfterUpdate(
+        @NonNull String key,
```

Comment:
what is the key ? jurisdiction ?

Reply:
we will use  jurisdiction  but it can be anything.

Reply:
changed